README.md
# Aerospike Express Session Store [![travis][travis-image]][travis-url] [![codecov][codecov-image]][codecov-url] [![npm][npm-image]][npm-url]

[travis-image]: https://travis-ci.org/aerospike/aerospike-session-store-expressjs.svg?branch=master
[travis-url]: https://travis-ci.org/aerospike/aerospike-session-store-expressjs
[codecov-image]: https://codecov.io/gh/aerospike/aerospike-session-store-expressjs/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/gh/aerospike/aerospike-session-store-expressjs
[npm-image]: https://img.shields.io/npm/v/aerospike-session-store.svg
[npm-url]: https://www.npmjs.com/package/aerospike-session-store

The Aerospike Express Session Store is an implementation of the Express.js
session store that uses Aerospike as the persistence layer for sessions. The
session middleware for Express is provided by the
[session-express](https://github.com/expressjs/session) module. Aerospike DB is
a high-performance NoSQL key-value store:
[www.aerospike.com](http://www.aerospike.com/).

## Installation

Via npm:

```bash
$ npm install aerospike-session-store
```
## Usage

### Initialization

Pass the `express-session` store into `aerospike-session-store` to create an
`AerospikeStore` constructor. Then use that constructor to create a new store
instance:

```js
const session = require('express-session')
const AerospikeStore = require('aerospike-session-store')(session)

var app = express()
app.use(session({
  secret: '123456789QWERTY',
  store: new AerospikeStore({
    namespace: 'express',
    set: 'session',
    ttl: 86400, // 1 day
    hosts: '10.0.0.1:3000,10.0.0.2:3000'
  }),
  resave: false,
  saveUninitialized: false
}))
```

### Options

The session store requires an Aerospike Client instance to connect to the DB
cluster. An existing client instance can be passed in using the `client`
option. Otherwise, the session store will create it's own client instance. The
Aerospike session store can be initialized with a number of optional
parameters:

* `client` - An existing Aerospike client instance that the session
  store should use instead of creating it's own instance.
* `namespace` - The Aerospike namespace to be used for session storage. (default: 'test')
* `set` - The Aerospike set name used when creating session records. (default: 'express-session')
* `ttl` - Time-to-live in seconds for the session records created in the
  Aerospike db. If not specified, the ttl will be determined based on the
  `maxAge` of the session cookie, if any. Set `ttl` to zero to disable usage of
  ttl. However, note that a default ttl at the namespace level might still apply.
* `mapper` - A custom data mapper to convert session objects to/from a
  format suitable for storage in an Aerospike record. By default, the `JSON`
  module is used to serialize session objects to/from JSON format.

Additional options are passed on to the Aerospike client when creating a new
client instance (i.e. unless the `client` paramter is used.) Please refer to
the client's [API
documentation](http://www.aerospike.com/apidocs/nodejs/Config.html) for a
detailed list of supported paramters.

## License

The Aerospike Express Session Store is made availabled under the terms of the
Apache License, Version 2, as stated in the file LICENSE.

Individual files may be made available under their own specific license, all
compatible with Apache License, Version 2. Please see individual files for
details.

History.md
v0.4.0 / 2017-10-06
===================

* **New Features**
  * Subscribe to client connect/disconnect events [PR #9](https://github.com/aerospike/aerospike-session-store-expressjs/pull/9)
  * BREAKING: Replace serializer with data mapper [PR #10](https://github.com/aerospike/aerospike-session-store-expressjs/pull/10)

v0.3.0 / 2017-10-05
===================

* **New Features**
  * Add AerospikeStore#clear method to delete all sessions
  * Add AerospikeStore#close method to close Aerospike client connection

* **Updates**
  * Update Aerospike client to v3

v0.2.0 / 2016-09-14
===================

* [#5](https://github.com/aerospike/aerospike-session-store-expressjs/pull/5) remove express dependency - Thanks to [@blitzprog](https://github.com/blitzprog)!

v0.1.0 / 2016-07-11
===================

* Update Aerospike client to v2
* Initial release to npmjs.org


test/index.js
var assert = require("assert");
var should = require("should");
var express = require('express');
var session = require('express-session');
var aerospikeStore = require('../lib/aerospike-session-store')(session);

var AeroSpikeStore = new aerospikeStore({
  ttl: 86400,
  hosts: ['127.0.0.1:3000'],
  prefix: 'sess:',
  ns: 'test'
});

var app = express();
app.use(session({
  secret: '123456789QWERTY',
  store: AeroSpikeStore,
  resave: false,
  saveUninitialized: false
}));


describe('AeroSpikeStore', function(){
  describe('set()', function(){
    it('should set session when no error', function(done){
     	 AeroSpikeStore.set('session1',{cookie:{maxAge:2000},name:'firstname lastname'}, function(err, ok){
          if(err) throw err;
     		done();
     	});
    })
  })
  
  describe('length()', function(){
  	it('should return 1 if session is present', function(done){
  		AeroSpikeStore.length(function(length){
        assert.equal(1, length);
        done();
      });
  	})
  })

  describe('get()', function(){
  	it('should return session if session is present', function(done){
  		AeroSpikeStore.get('session1', function(err, ok){
  			assert.ok(err.code === 0, 'get error!!!!');
  			done(); 		
  		});
  	})
    it('should throw error in case session is not found',function(done){
      assert.throws(AeroSpikeStore.get('session2', function(err, ok){
        if(err) throw err;
        done();
      })); 
    })
	it('should return session if session has not expired', function(done){
  		AeroSpikeStore.get('session1', function(err, ok){
  			assert.ok(ok.cookie.maxAge === 2000, 'get maxAge error!!!!');
  			done();		
		});
	})
  it('should throw error in case session has expired',function(done){
      AeroSpikeStore.get('session1', function(err, ok){
        assert.notEqual(ok.cookie.maxAge,3000);
        done();
      }); 
    })
	it('should not return error if session value is matching', function(done){
		AeroSpikeStore.get('session1', function(err, ok){
			assert.ok(ok.name === 'last lastname', 'get name error!!!');
			done();		
	});
	})
  it('should throw error in case session value is not matching',function(done){
    AeroSpikeStore.get('session1', function(err, ok){
      assert.notEqual(ok.name,'firstname m. lastname', 'get name error!!!');
      done();
    }); 
  })
})

  describe('destroy()', function(){
  	it('should not throw error if able to destroy session', function(done){
  		AeroSpikeStore.destroy('session1', function(err){
  			if(err) throw err;
     		done();
  		});
      
  	})
  })

  describe('length()', function(){
    it('should return 0 if session is destroyed', function(done){
      AeroSpikeStore.length(function(length){
        assert.equal(0, length);
        done();
      });
    })
  })
});

test/aerospike_store_test.js
// *****************************************************************************
// Copyright 2016-2017 Aerospike, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License")
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// *****************************************************************************

'use strict'

const Promise = require('bluebird')
const test = require('blue-tape')
const session = require('express-session')
const AerospikeStore = require('../')(session)
const Aerospike = require('aerospike')

function delay (ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

function lifecycleTest (store, t) {
  Promise.promisifyAll(store)

  const session = { cookie: { maxAge: 2000 }, name: 'jan' }
  return store.setAsync('sid', session)
    .then(() => t.pass('#set() ok'))
    .then(() => store.getAsync('sid'))
    .then(data => t.deepEqual({ cookie: { maxAge: 2000 }, name: 'jan' }, data, '#get() ok'))
    .then(() => store.touchAsync('sid', session))
    .then(() => t.pass('#touch() ok'))
    .then(() => store.destroyAsync('sid'))
    .then(() => store.getAsync('sid'))
    .then(data => t.equal(undefined, data, '#destroy() ok'))
    .then(() => store.close(false))
}

test('defaults', function (t) {
  const store = new AerospikeStore()
  t.equal(store.as_namespace, 'test', 'sets default namespace to test')
  t.equal(store.as_set, 'express-session', 'sets default set name to express-session')
  t.notOk(store.ttl, 'ttl not set')
  t.ok(store.client, 'creates client')

  store.client.close(false)
  t.end()
})

test('basic', function (t) {
  t.throws(AerospikeStore, TypeError, 'constructor not callable as function')
  const store = new AerospikeStore()
  return lifecycleTest(store, t)
})

test('clear', function (t) {
  const store = new AerospikeStore()
  Promise.promisifyAll(store)

  return store.setAsync('sess1', { name: 'jan' })
    .then(() => store.clearAsync())
    .then(() => delay(200))
    .then(() => store.getAsync('sess1'))
    .then(session => t.equal(session, undefined, 'all sessions cleared'))
    .then(() => store.close(false))
})

test('existing client', function (t) {
  return Aerospike.connect()
    .then(function (client) {
      const store = new AerospikeStore({ client: client })
      return lifecycleTest(store, t)
    })
})

test('options', function (t) {
  const store = new AerospikeStore({
    namespace: 'express',
    set: 'session',
    ttl: 3600
  })

  t.equal(store.as_namespace, 'express', 'uses provided namespace')
  t.equal(store.as_set, 'session', 'uses provided set name')
  t.equal(store.ttl, 3600, 'sets ttl')

  return lifecycleTest(store, t)
})

test('failed connection', function (t) {
  const store = new AerospikeStore({ hosts: '127.0.0.1:3333', connTimeoutMs: 500 })
  Promise.promisifyAll(store)

  return store.setAsync('sid', { cookie: { maxAge: 2000 }, name: 'jan' })
    .catch(function (err) {
      t.ok(/Not connected/.test(err.message), 'failed connection')
      store.close(false)
    })
})

test('serializer', function (t) {
  const serializer = {
    stringify: (value, replacer, space) => 'XXX' + JSON.stringify(value, replacer, space),
    parse: str => {
      t.ok(str.match(/^XXX/))
      return JSON.parse(str.substring(3))
    }
  }
  t.equal(serializer.stringify('UnitTest'), 'XXX"UnitTest"')
  t.equal(serializer.parse(serializer.stringify('UnitTest')), 'UnitTest')

  const store = new AerospikeStore({ serializer: serializer })
  return lifecycleTest(store, t)
})

test('serializer error', function (t) {
  const a = {}
  const b = {}
  // create circular reference to cause serialization error
  a.b = b
  b.a = a

  const store = new AerospikeStore()
  store.set('sid', a, error => {
    t.assert(error && error.message.match(/circular/), 'serialization error is passed to callback')
    store.close(false)
    t.end()
  })
})

test('after', function (t) {
  Aerospike.releaseEventLoop()
  t.end()
})


lib/data_mapper.js
// *****************************************************************************
// Copyright 2016-2017 Aerospike, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License")
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// *****************************************************************************

'use strict'

class DataMapper {
  /**
   * Converts a session object into a database record.
   *
   * @param {Object} session - session object
   * @returns {Object} Aerospike database record
   */
  toRecord (session) {
    return {
      session: JSON.stringify(session)
    }
  }

  /**
   * Converts a database record into a session object.
   *
   * @param {Object} record - Aerospike record bins
   * @param {Object} session object
   */
  fromRecord (record) {
    const sessionBin = record.session
    if (!sessionBin) return null
    return JSON.parse(sessionBin)
  }
}

module.exports = DataMapper


examples/express.js
const express = require('express')
const session = require('express-session')
const AerospikeStore = require('aerospike-session-store')(session)

var app = express()
app.use(session({
  secret: '123456789QWERTY',
  store: new AerospikeStore({
    namespace: 'express',
    set: 'session',
    ttl: 86400, // 1 day
    hosts: process.env.AEROSPIKE_HOSTS
  }),
  resave: false,
  saveUninitialized: false
}))
app.get('/', (req, res) => {
  var sess = req.session
  if (sess.views) {
    sess.views++
    res.setHeader('Content-Type', 'text/html')
    res.write('<p>views: ' + sess.views + '</p>')
    res.end()
  } else {
    console.log('initializing new session: ', req.sessionID)
    sess.views = 1
    res.end('welcome to the session demo. refresh!')
  }
  console.log('[%s] views=%s', req.sessionID, sess.views)
})
app.listen(3000, () => {
  console.log('Session store example listening at http://localhost:3000/.')
})


lib/aerospike_store.js
// *****************************************************************************
// Copyright 2016-2017 Aerospike, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License")
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// *****************************************************************************

'use strict'

const Aerospike = require('aerospike')

const debug = require('debug')('session:aerospike')
const util = require('util')

module.exports = function (session) {
  const Store = session.Store

  /**
   * Return the `AerospikeStore` extending `express`'s session Store.
   *
   * @param {Object} express session
   * @return {Function}
   * @public
   */
  function AerospikeStore (options) {
    if (!(this instanceof AerospikeStore)) {
      throw new TypeError('Cannot call AerospikeStore constructor as a function')
    }

    options = options || {}
    Store.call(this, options)

    this.as_namespace = options.namespace || 'test'
    this.as_set = options.set || 'express-session'
    this.ttl = options.ttl
    this.serializer = options.serializer || JSON

    if (options.client) {
      this.client = options.client
    } else {
      this.client = Aerospike.client(options)
      this.client.connect(error => {
        if (error) {
          debug('ERROR - %s', error.message)
          process.nextTick(() => this._onDisconnected())
        } else {
          process.nextTick(() => this._onConnected())
        }
      })
    }

    this.connected = false
    this.client.on('nodeAdded', () => this._onConnected())
    this.client.on('disconnected', () => this._onDisconnected())
  }

  util.inherits(AerospikeStore, Store)

  AerospikeStore.prototype._onConnected = function () {
    if (this.connected) return
    debug('CONNECT')
    this.connected = true
    this.emit('connect')
  }

  AerospikeStore.prototype._onDisconnected = function () {
    if (!this.connected) return
    debug('DISCONNECT')
    this.connected = false
    this.emit('disconnect')
  }

  AerospikeStore.prototype._key = function (sid) {
    return new Aerospike.Key(this.as_namespace, this.as_set, sid)
  }

  /**
   * Determines the time-to-live (TTL) to use when writing a record for
   * session.set() or session.touch(). The default ttl can be set by passing
   * the ttl option when creating a new store instance. Otherwise, the maxAge
   * of the session cookie will be used to determine ttl. If ttl is zero or
   * negative, ttl will not be used. However, note that the Aerospike server
   * might still have a default ttl at the namespace level.
   *
   * @param {Object} express session
   * @private
   */
  AerospikeStore.prototype._ttl = function (session) {
    let ttl = null
    if (typeof this.ttl === 'number') {
      ttl = this.ttl
    } else {
      const maxAge = session.cookie && session.cookie.maxAge
      if (typeof maxAge === 'number') {
        ttl = Math.floor(maxAge / 1000)
      }
    }
    if (ttl <= 0) {
      ttl = null
    }
    return ttl
  }

  /**
   * Get a session from the store given a session ID `sid`.
   *
   * @param {String} sid - Session ID
   * @param {Function} callback - Callback function
   * @public
   */
  AerospikeStore.prototype.get = function (sid, callback) {
    const key = this._key(sid)
    debug('GET "%s"', sid)
    this.client.get(key, (err, record) => {
      if (err) {
        switch (err.code) {
          case Aerospike.status.AEROSPIKE_ERR_RECORD_NOT_FOUND:
            return callback()
          default:
            return callback(err)
        }
      }

      const bins = record.bins
      const sessionBin = bins.session
      if (!sessionBin) {
        return callback()
      }

      try {
        const session = this.serializer.parse(sessionBin)
        return callback(null, session)
      } catch (err) {
        return callback(err)
      }
    })
  }

  /**
   * Remove a session from the store given a session ID `sid`.
   *
   * @param {String} sid - Session ID
   * @param {Function} callback - Callback function
   */
  AerospikeStore.prototype.destroy = function (sid, callback) {
    const key = this._key(sid)
    debug('REMOVE "%s"', sid)
    this.client.remove(key, error => callback(error))
  }

  /**
   * Upsert a session into the store given a session ID `sid` and session
   * object `session`.
   *
   * @param {String} sid - Session ID
   * @param {Object} session - Session object
   * @param {Function} callback - Callback function
   * @public
   */
  AerospikeStore.prototype.set = function (sid, session, callback) {
    try {
      const key = this._key(sid)
      const bins = { session: this.serializer.stringify(session) }
      const meta = { ttl: this._ttl(session) }
      debug('PUT "%s", ttl: %s', sid, meta.ttl)
      this.client.put(key, bins, meta, error => callback(error))
    } catch (error) {
      process.nextTick(callback, error)
    }
  }

  /**
   * Refresh the time-to-live for the session with the given session ID `sid`.
   *
   * @param {String} sid - Session ID
   * @param {Object} session - Session object
   * @param {Function} callback - Callback function
   * @public
   */
  AerospikeStore.prototype.touch = function (sid, session, callback) {
    const key = this._key(sid)
    const ttl = this._ttl(session)
    if (ttl) {
      debug('TOUCH "%s" %s', sid, ttl)
      const ops = [ Aerospike.operations.touch(ttl) ]
      this.client.operate(key, ops, error => callback(error))
    } else {
      process.nextTick(callback)
    }
  }

  /**
   * Delete all sessions from the store.
   *
   * @param {Function} callback - Callback function
   * @public
   */
  AerospikeStore.prototype.clear = function (callback) {
    debug('CLEAR')
    this.client.truncate(this.as_namespace, this.as_set, error => callback(error))
  }

  /**
   * Closes the Aerospike client connection.
   *
   * @param {boolean} [releaseEventLoop=true] - Whether to release the event
   * loop handle after the client is closed.
   *
   * @see http://www.aerospike.com/apidocs/nodejs/module-aerospike.html#.releaseEventLoop
   *
   * @public
   */
  AerospikeStore.prototype.close = function (releaseEventLoop) {
    debug('CLOSE - release event loop: %s', releaseEventLoop)
    this.client.close(releaseEventLoop)
    this._onDisconnected()
  }

  return AerospikeStore
}


.travis/start_cluster.sh
#!/bin/bash
################################################################################
# Copyright 2013-2017 Aerospike, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

nodes=${1:-1}
home=$(pwd)

function wait_for_node {
  log=$1
  if [ ! -f $log ]
  then
    echo "A log file does not exist at $(pwd)/${log}"
    exit 1
  fi

  i=0
  while [ $i -le 12 ]
  do
    sleep 1
    grep -i "there will be cake" ${log}
    if [ $? == 0 ]; then
      return 0
    else
      i=$(($i + 1))
      echo -n "."
    fi
  done
  echo "The cake is a lie!"
  tail -n 1000 ${log}
  exit 2
}

function start_server {
  instance=$1
  dir="instance${instance}"
  port=$((2900 + 100 * $instance))
  echo "## Starting Aerospike instance ${instance}/${nodes} on port ${port}"
  mkdir ${dir}
  ./bin/aerospike init --home ${dir} --instance ${instance} --service-port ${port}
  cd ${dir}
  ./bin/aerospike start
  wait_for_node "var/log/aerospike.log"
  cd ..
}

function install_server {
  wget -nv -O aerospike-server.tgz http://aerospike.com/download/server/latest/artifact/tgz
  tar xzf aerospike-server.tgz
  cp -f .travis/aerospike.conf ./aerospike-server/share/etc
  cd aerospike-server
  sed -i -e 's/\${me}/"root"/' share/libexec/aerospike-start
  sed -i -e 's/set_shmmax$/#set_shmmax/' share/libexec/aerospike-start
  sed -i -e 's/set_shmall$/#set_shmall/' share/libexec/aerospike-start
  sed -i -e 's/set_socket_buffer_limits$/#set_socket_buffer_limits/' share/libexec/aerospike-start
  sed -i -e 's/ulimit/#ulimit/' share/libexec/aerospike-start
}

echo "## Fetching Aerospike server distribution"
install_server
for ((node = 1; node <= nodes; node++))
do
  start_server $node
done
cd ${pwd}


benchmark/aerospikebench.js
const session = require('express-session')
const AerospikeStore = require('../')(session)

var store = new AerospikeStore({ ttl: 3600 })

var sessionsToCreate = 100000
var sessionsCreated = 0
var inFlight = 0
var maxConcurrent = 300

var payload = { cookie: { maxAge: 2000 }, name: 'sample name' }

var task = function (err, init) {
  if (err) throw err
  if (!init) inFlight--
  if (sessionsCreated < sessionsToCreate && inFlight < maxConcurrent) {
    var sessionID = 'test' + (sessionsCreated + 1)
    store.set(sessionID, payload, task)
    sessionsCreated++
    inFlight++
  } else if (sessionsCreated === sessionsToCreate && inFlight === 0) {
    console.timeEnd('bench ' + sessionsToCreate)
    store.close()
  }
}

console.time('bench ' + sessionsToCreate)
for (var i = 0; i < maxConcurrent; i++) {
  task(null, true)
}


lib/aerospike-session-store.js
/*
* aerospike session store module
* saves session information in aerospike
*/

/* 
* dependencies
*/
var _ = require('lodash');
//var crypto = require('crypto');
var util = require('util');
var aerospike = require('aerospike');



/**
 * Default options
 */
var _default_port = 3000;

var defaultOptions = {
  hosts: ['127.0.0.1:3000'],
  log_level: '5',
  time_out: 10,
  prefix: 'sess:',
  stringify: true,
  hash: false,
  ns: 'test',
  st: 'store',
  ttl:  60 * 60 * 24 * 14, // 14 days
};


var defaultHashOptions = {
  salt: '123456789QWERTY',
  algorigthm: 'sha1'
};


var defaultSerializationOptions = {
  serialize: function(session){
    var obj = {};
    for(var prop in session){
      if(prop === 'cookie'){
        obj.cookie = session.cookie.toJSON ? session.cookie.toJSON() : session.cookie;
      } else {
        obj[prop] = session[prop];
      }
    }
    return obj;
  },
  unserialize: _.identity
};

var stringifySerializationOptions = {
  serialize: JSON.stringify,
  unserialize: JSON.parse
};


/*
* aerospike-session-store
* returns AerospikeSessionStore object
*/

module.exports = function(session){
	 
	var Store = session.Store;
	


/*
* initialize aerospike-session-store with passed options
* @param {object} options
*/
	function AerospikeSessionStore(options){
		var self = this;
		options = _.defaults(options || {}, defaultOptions);
		Store.call(self, options);
		//Store = session.Store;

		self.prefix = options.prefix;
    	self.ns = options.ns;
    	self.st = options.st;
    	self.ttl = options.ttl;
    	//console.log(options.ns);
		if(options.hosts){
			var hosts = options.hosts.map(function(host){
				var _val = host.split(':');
				var _host = _val[0];
				  var _port = ~~_val[1] || _default_port
				return { addr:_host,port:parseInt(_port)};
			});
		}

		/* connection to aerospike */
		aerospike.client({
			hosts:hosts,
			log:{
				level: options.log_level
			},
			policies: {
				timeout: options.time_out
			}
		}).connect(function(err,client){
			if(err.code != aerospike.status.AEROSPIKE_OK){
				console.error('Aerospike server connection Error : %j', err);
			} else {
				self.client = client;
			}
		});
		
		this.getSessionId = function(sid){
      		if(options.hash){
        		return crypto.createHash(options.hash.algorithm).update(options.hash.salt + this.prefix + sid).digest('hex');
      		} else {
        		return this.prefix + sid;
      		}
    	}

	};

/* inherit from Store */
	util.inherits(AerospikeSessionStore, Store);

/*
* save a particular session to aerospike
* @param {string} sid
* @param {session} session
* @param {function} callback
*/

	AerospikeSessionStore.prototype.set = function(sid, session, callback){
		if(!callback) callback = _.noop;
		//console.log(session);
		sid = this.getSessionId(sid);
		var maxAge = session.cookie.maxAge;
		var ttl = this.ttl;
		var key = aerospike.key(this.ns, this.st, sid);
		//console.log(key);
		ttl = ttl || (typeof maxAge === 'number' ? maxAge/1000 | 0 : oneDay);
		var meta = {gen:0, ttl:ttl};
		var sessionString = JSON.stringify(session);
		this.client.put(key, {sessionString: sessionString}, meta, function(err, key){
			//console.log(err);
			if(err.code != aerospike.status.AEROSPIKE_OK){
				
				callback(err);
			}else {
				
				callback(null, key);
			}
		});
		//return(sid+ ' - ' + maxAge + ' ' +ttl + ' '+ key);
	};

/*
* destroy a particular session value
* @param {string} sid
* @param {function} callback
*/
	AerospikeSessionStore.prototype.get = function(sid, callback){
		if(!callback) callback = _.noop;
		sid = this.getSessionId(sid);
		var key = aerospike.key(this.ns, this.st, sid);
		//console.log(key);
		this.client.get(key, function(err, res, meta){
			if(err.code	 === aerospike.status.AEROSPIKE_OK){
				var session;
				if(typeof res.sessionString === 'undefined'){
					session = res;
				} else{
					var sessionString = res.sessionString || '{}';
					session = JSON.parse(sessionString);
				}
				//console.log(sessionString);
				return callback({code: 0}, session);
			} else if(err.code === aerospike.status.AEROSPIKE_ERR_RECORD_NOT_FOUND){
				return callback();
			} else {
				return callback(err);
			}
		});
	},

/*
* calculate session length
* @param {function} callback
*/

	AerospikeSessionStore.prototype.length = function(callback){
		if(!callback) callback = _.noop;
		var key = aerospike.key(this.ns, this.st);
		var infocommand = 'sets/'+this.ns+'/'+this.st;
		
		this.client.info(infocommand,function(err, res, key){
				//console.log(res);
				var count_user = res.split(':');
				count_user = count_user[0].split('=');
				var numSession = 0;
				if(parseInt(count_user[1])>0){
						numSession = parseInt(count_user[1]);
					}
					//console.log(numSession);
				return callback(numSession);
		});
	};

/*
* destroy a particular session
* @param {string} sid
* @param {function} callback
*/

	AerospikeSessionStore.prototype.destroy = function(sid, callback){
		if(!callback) callback = _.noop;
		sid = this.getSessionId(sid);
		var key = aerospike.key(this.ns, this.st, sid);
		this.client.remove(key, function(err, key){
			if(err.code != aerospike.status.AEROSPIKE_OK){
				//console.log('error %s', err.message);
				callback(err);
			}else {
				//console.log('session destroyed')
				callback(null, key);
			}
		});


	};

return AerospikeSessionStore;

};

benchmark/benchmarktest.js
var Benchmark = require('benchmark');
// basic usage (the `new` operator is optional)
// or options only
var assert = require("assert");
var should = require("should");
var express = require('express');
var session = require('express-session');
var aerospikeStore = require('../lib/aerospike-session-store')(session);

var AeroSpikeStore = new aerospikeStore({
  ttl: 86400,
  hosts: ['127.0.0.1:3000'],
  prefix: 'sess:',
  ns: 'test'
});

var app = express();
app.use(session({
  secret: '123456789QWERTY',
  store: AeroSpikeStore,
  resave: false,
  saveUninitialized: false
}));

var suite = new Benchmark.Suite;
 
// add tests 
suite.add('AeroSpikeStore#set', function() {
  
  AeroSpikeStore.set('pranav',{cookie:{maxAge:2000},name:'nilesh gawde'}, function(err, ok){
          if(err) throw err;
         // done();
        });
    ///o/.test('Hello World!');
})
.add('AeroSpikeStore#get', function() {
    AeroSpikeStore.get('pranav', function(err, ok){  
        assert.ok(err.code === 0, 'get error!!!!');
    });
})
.add('AeroSpikeStore#length', function() {
    AeroSpikeStore.length(function(length){return length;});
})
.add('AeroSpikeStore#destroy', function() {
    AeroSpikeStore.destroy('pranav', function(err){
       // if(err.message) throw err;
       // done();
      });
})
// add listeners 
.on('cycle', function(event) {
  console.log(String(event.target));
})
.on('complete', function() {
  console.log('Fastest is ' + this.filter('fastest').pluck('name'));
})
// run async 
.run({ 'async': true });
 
// logs: 
// > RegExp#test x 4,161,532 +-0.99% (59 cycles) 
// > String#indexOf x 6,139,623 +-1.00% (131 cycles) 
// > Fastest is String#indexOf 